source/posix/system_info.c

```diff
@@ -445,3 +450,118 @@ enum aws_platform_os aws_get_platform_build_os(void) {
     return AWS_PLATFORM_OS_UNIX;
 }
 #endif /* AWS_OS_APPLE */
+
+/** Static variables to store the last CPU usage call. */
+static uint64_t s_cpu_last_total_user = 0;
+static uint64_t s_cpu_last_total_user_low = 0;
+static uint64_t s_cpu_last_total_system = 0;
+static uint64_t s_cpu_last_total_idle = 0;
+
+static void s_getCurrentCpuUsage(
+    uint64_t *total_user,
+    uint64_t *total_user_low,
+    uint64_t *total_system,
+    uint64_t *total_idle) {
+    *total_user = 0;     // prevent warnings over unused parameter on Mac
+    *total_user_low = 0; // prevent warnings over unused parameter on Mac
+    *total_system = 0;   // prevent warnings over unused parameter on Mac
+    *total_idle = 0;     // prevent warnings over unused parameter on Mac
+
+// Get the CPU usage from Linux
+#if defined(__linux__) || defined(__unix__)
+    FILE *file;
+    int matched_results;
+    file = fopen("/proc/stat", "r");
+    matched_results = fscanf(
+        file,
+        "cpu %llu %llu %llu %llu",
+        (long long unsigned int *)total_user,
+        (long long unsigned int *)total_user_low,
+        (long long unsigned int *)total_system,
+        (long long unsigned int *)total_idle);
+    fclose(file);
+    if (matched_results == EOF || matched_results != 4) {
+        aws_raise_error(AWS_ERROR_INVALID_STATE);
+    }
+#endif
+}
+
+int aws_get_cpu_usage(double *output) {
+// Get the CPU usage from Linux
+#if defined(__linux__) || defined(__unix__)
+    int return_result = AWS_OP_ERR;
+    uint64_t total_user, total_user_low, total_system, total_idle, total;
+    s_getCurrentCpuUsage(&total_user, &total_user_low, &total_system, &total_idle);
+    double percent;
+
+    // Overflow detection
+    if (total_user < s_cpu_last_total_user || total_user_low < s_cpu_last_total_user_low ||
+        total_system < s_cpu_last_total_system || total_idle < s_cpu_last_total_idle) {
+        *output = 0;
+    } else {
+        total = (total_user - s_cpu_last_total_user) + (total_user_low - s_cpu_last_total_user_low) +
+                (total_system - s_cpu_last_total_system);
+        percent = total;
+        total += total_idle - s_cpu_last_total_idle;
+        percent = (percent / total) * 100;
+
+        // If percent is negative, then there was an error (overflow?)
```

Comment:
Would be better to detect previous failures directly using the appropriate math functions (aws_*_u64_checked)